airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,309 @@
+      boolean hasState = ctx.selectFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              streamName != null ? STATE.STREAM_NAME.eq(streamName) : STATE.STREAM_NAME.isNull(),
+              namespace != null ? STATE.NAMESPACE.eq(namespace) : STATE.NAMESPACE.isNull())
+          .fetch().isNotEmpty();
+
+      final JSONB jsonbState = JSONB.valueOf(Jsons.serialize(state));
+      final OffsetDateTime now = OffsetDateTime.now();
+
+      if (!hasState) {
+        ctx.insertInto(STATE)
+            .columns(
+                STATE.ID,
+                STATE.CREATED_AT,
+                STATE.UPDATED_AT,
+                STATE.CONNECTION_ID,
+                STATE.STREAM_NAME,
+                STATE.NAMESPACE,
+                STATE.STATE_,
+                STATE.TYPE)
+            .values(
+                UUID.randomUUID(),
+                now,
+                now,
+                connectionId,
+                streamName,
+                namespace,
+                jsonbState,
+                convertStateType(stateType))
+            .execute();
+
+      } else {
+        ctx.update(STATE)
+            .set(STATE.UPDATED_AT, now)
+            .set(STATE.STATE_, jsonbState)
+            .where(
+                STATE.CONNECTION_ID.eq(connectionId),
+                streamName != null ? STATE.STREAM_NAME.eq(streamName) : STATE.STREAM_NAME.isNull(),
+                namespace != null ? STATE.NAMESPACE.eq(namespace) : STATE.NAMESPACE.isNull())
+            .execute();
+      }
+
+    } else {
+      // If the state is null, we remove the state instead of keeping a null row
+      ctx.deleteFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              streamName != null ? STATE.STREAM_NAME.eq(streamName) : STATE.STREAM_NAME.isNull(),
+              namespace != null ? STATE.NAMESPACE.eq(namespace) : STATE.NAMESPACE.isNull())
+          .execute();
+    }
+  }
+
+  /**
+   * Get the StateType for a given list of StateRecords
+   *
+   * @param connectionId The connectionId of the records, used to add more debugging context if an
+   *        error is detected
+   * @param records The list of StateRecords to process, must not be empty
+   * @return the StateType of the records
+   * @throws IOException If StateRecords have inconsistent types
+   */
+  private io.airbyte.db.instance.configs.jooq.generated.enums.StateType getStateType(final UUID connectionId, final List<StateRecord> records)
+      throws IOException {
+    final List<io.airbyte.db.instance.configs.jooq.generated.enums.StateType> types = records.stream().map(r -> r.type).distinct().toList();
+    if (types.size() == 1) {
+      return types.get(0);
+    }
+
+    throw new IOException("Inconsistent StateTypes for connectionId " + connectionId +
+        " (" + String.join(", ", types.stream().map(io.airbyte.db.instance.configs.jooq.generated.enums.StateType::getLiteral).toList()) + ")");
```

Comment:
Nit/don't have to modify: I think that `map(stateType -> stateType.getLiteral())` will be easier to read :D

Reply:
I was not convinced due to the full path to the class as well. your comment convinced me 😂